articles/modules/ROOT/pages/protecting-against-cypher-injection.adoc

```diff
@@ -38,130 +38,466 @@ Result result = session.run(queryString);
+`LilRob' OR 1=1 WITH true AS hacked MATCH (all) DETACH DELETE all; //`
 
-== When you MUST append into a query string, sanitize your inputs
+As the username is received from the user directly, our application sets it user a parameter.
 
-There are some cases where we can't pass a parameter into the procedure call, or a query, and string appending is the only option.
+`:param username => "LilRob' OR 1=1 WITH true AS hacked MATCH (all) DETACH DELETE all; //"`
 
-For example, there are some things in Cypher that cannot be parameterized, such as node labels and relation types.
-There are some APOC Procs that can help (and should be used if so), but aside from these, the only option may be to append into a query string.
+[source,cypher]
+----
+CREATE (u:User) SET u.username = $username;
+----
+
+Now that an account is made, Little Robby Tables logs in and goes to the settings to change his username.
+The database retrieves his current username and uses client-side string concatenation to build a query to update it.
+
+`query = "MATCH (u:User) WHERE u.username = '" + username + "' SET u.username = $newUsername;"`
+
+This query is executed as:
+
+[source,cypher]
+----
+MATCH (u:User) WHERE u.username = 'LilRob' OR 1=1 WITH true AS hacked MATCH (all) DETACH DELETE all; //' SET u.username = $newUsername;
+----
 
-In these cases, it is extremely important to sanitize your inputs, removing quote or delimiter characters (depending on their context of use) that would allow input to break out of the context within which you're trying to use it.
+The malicious code is now run, and all users are deleted!
+This is why sanitization should continue to be used, even when the input doesn't appear to be coming directly from a user.
 
-In these cases it is better to sanitize input in your own code at the client level, as there are many utilities across various languages for input sanitization, and it makes sense to address it at that level rather than lower down at the database itself.
+== Role Based Privileges
+
+=== Principle of Least Privilege
+
+The principle of least privilege is the idea that a program or user should have the bare minimum of privileges needed to perform their function.
+This means that, for example, if your application is only reading data, then it should have read only access to that data.
+The benefit of this is that in the case of a Cypher Injection attack, an injected query is not able to manipulate the data,
+as the role executing the hijacked query is limited in its functionality.
+With Neo4j, a range of fine-grained access control is available, which can add another layer of protection in case of an injection attack.
+See https://neo4j.com/docs/operations-manual/current/authentication-authorization/access-control/[here] for more information on role based privileges in Neo4j.
```

Comment:
Should we: 
* mention this is only possible in the enterprise version
* explicitely mention restricting procedures that are not needed (could be about whitelisting procedures from a jar as well)

Reply:
There are still roles available in community I thought, e.g read? or do I have that wrong?

I can try to add something about allowlists :)

Reply:
AFAIK, in community edition these permissions are not available, and every user is essentially acting as an admin (eg they can do everything).

Reply:
I don't know if we need to focus on any particular setting (eg procedure.allowlist) but I think we should perhaps say that principle of least privilege can be controlled both via RBAC and configuration settings.

Reply:
Wow, I didn't realise community was so limited! Will mention enterprise then :) 

I like that suggestion Daniel! thanks, will generalize it a little :D